craft_parts/executor/part_handler.py

```diff
@@ -245,6 +244,131 @@ def _run_step(
 
         return step_handler.run_builtin()
 
+    def _update_action(self, action: Action, *, step_info: StepInfo) -> None:
+        handler: Callable[[StepInfo], None]
+
+        if action.step == Step.PULL:
+            handler = self._update_pull
+        elif action.step == Step.BUILD:
+            handler = self._update_build
+        else:
+            step_name = action.step.name.lower()
+            raise errors.InvalidAction(
+                f"cannot update step {step_name!r} of {self._part.name!r}."
+            )
+
+        callbacks.run_pre_step(step_info)
+        handler(step_info)
+        state_file = states.state_file_path(self._part, action.step)
+        state_file.touch()
+        callbacks.run_post_step(step_info)
+
+    def _update_pull(self, step_info: StepInfo) -> None:
+        self._make_dirs()
+
+        # if there's an override-pull scriptlet, execute it instead
+        if self._part.spec.override_pull:
+            self._run_step(
+                step_info=step_info,
+                scriptlet_name="override-pull",
+                work_dir=self._part.part_src_dir,
+            )
+            return
+
+        # consistency check
+        if not self._source_handler:
+            raise RuntimeError(
+                f"Update requested on part {self._part.name!r} without "
+                f"a source handler."
+            )
+
+        # the update action is sequenced only if an update is required and the
+        # source knows how to update
+        state_file = states.state_file_path(self._part, step_info.step)
+        self._source_handler.check_if_outdated(str(state_file))
+        self._source_handler.update()
+
+    def _update_build(self, step_info: StepInfo) -> None:
+        self._make_dirs()
+        self._unpack_stage_packages()
+        self._unpack_stage_snaps()
+
+        if not self._plugin.out_of_source_build:
+            # Use the local source to update. It's important to use
+            # file_utils.copy instead of link_or_copy, as the build process
+            # may modify these files
+            source = sources.LocalSource(
```

Comment:
Would it be reasonable to simply call the copy rather than attempt to use the local plugin?  Or is the `check_if_outdated()` call providing value?

Reply:
Yes, `check_if_outdated()` must be called before `update`. I'll add a note to the source code.